registry/mdns_registry.go

```diff
@@ -258,25 +281,68 @@ func (m *mdnsRegistry) Register(service *Service, opts ...RegisterOption) error
 			continue
 		}
 
-		e.id = node.Id
-		e.node = srv
-		entries = append(entries, e)
+		entries = append(entries, &mdnsEntry{id: node.Id, node: srv})
 	}
 
-	// save
-	m.services[service.Name] = entries
+	// save the mdns entry
+	m.domains[options.Domain][service.Name] = entries
+	m.Unlock()
+
+	// register in the global tld so it can be queried as one
+	if options.Domain != m.globalTLD {
+		srv := *service
+		srv.Nodes = nil
+
+		for _, n := range service.Nodes {
+			node := n
+
+			// set the original domain in node metadata
+			if node.Metadata == nil {
+				node.Metadata = map[string]string{"domain": options.Domain}
+			} else {
+				node.Metadata["domain"] = options.Domain
+			}
+
+			srv.Nodes = append(srv.Nodes, node)
+		}
+
+		if err := m.Register(service, append(opts, RegisterDomain(m.globalTLD))...); err != nil {
+			gerr = err
+		}
+	}
 
 	return gerr
 }
 
 func (m *mdnsRegistry) Deregister(service *Service, opts ...DeregisterOption) error {
+	// parse the options
+	var options DeregisterOptions
+	for _, o := range opts {
+		o(&options)
+	}
+	if len(options.Domain) == 0 {
+		options.Domain = m.defaultTLD
+	}
+
+	// register in the global tld
+	if options.Domain != m.globalTLD {
+		// no errors are returned by this function so we don't need to handle the error
+		defer m.Deregister(service, append(opts, DeregisterDomain(m.globalTLD))...)
```

Comment:
You're going to lose the error by doing the defer

Reply:
Yes, this was expected but because the function doesn't return an error, it's okay. I did add a comment for this. Would it be better to simply return m.Deregister everywhere return is called?

Reply:
I just looked at the function signature and saw it returned err (I missed your comment).